### PR TITLE
Enable secondary optimization of the preventive state

### PR DESCRIPTION
**Pull request by pet-mit**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
Only one preventive RAO

**What is the new behavior (if this is a feature change)?**
A second preventive RAO is possible to relieve curative constraints

### CODE REVIEW

File: sensitivity-analysis/src/main/java/com/farao_community/farao/sensitivity_analysis/SystematicSensitivityResult.java [hunk 1]
```diff
@@ -58,23 +59,29 @@
...
     private final Map<String, StateResult> contingencyResults = new HashMap<>();
+    private final Map<String, StateResult> contingencyResultsAfterCra = new HashMap<>();
 
     public SystematicSensitivityResult() {
         this.status = SensitivityComputationStatus.SUCCESS;
     }
 
     public SystematicSensitivityResult completeData(SensitivityAnalysisResult results) {
+        return completeData(results, false);
+    }
```
Comment (phiedw) [inline sensitivity-analysis/src/main/java/com/farao_community/farao/sensitivity_analysis/SystematicSensitivityResult.java:70]: i think having only the method where we explicitely need the afterCra argument would be safer than having both these methods

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTreeBloomer.java [hunk 2]
```diff
@@ -156,7 +156,11 @@ private boolean isNetworkActionCloseToLocations(NetworkAction networkAction, Set
     }
 
     private Set<Optional<Country>> getOptimizedMostLimitingElementLocation(Leaf leaf) {
-        FlowCnec cnec = leaf.getMostLimitingElements(1).get(0);
-        return cnec.getLocation(network);
+        List<FlowCnec> mostLimitingElement = leaf.getMostLimitingElements(1);
+        if (mostLimitingElement.isEmpty()) {
+            // can happen in curative perimeters with MNEC only
+            return new HashSet<>();
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTreeBloomer.java]: fixed

File: sensitivity-analysis/src/main/java/com/farao_community/farao/sensitivity_analysis/SystematicSensitivityResult.java (hunk 1 above)
Comment (pet-mit) [reply sensitivity-analysis/src/main/java/com/farao_community/farao/sensitivity_analysis/SystematicSensitivityResult.java:70]: ok

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTreeInput.java [hunk 3]
```diff
@@ -31,6 +33,9 @@
 
     private PrePerimeterResult prePerimeterOutput;
 
+    private Map<State, Set<NetworkAction>> appliedNetworkActions;
+    private Map<State, Map<RangeAction, Double>> appliedRemedialActionsSetpoints;
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTreeInput.java:37]: these are not needed actually, i forgot to remove them

File: sensitivity-analysis/src/main/java/com/farao_community/farao/sensitivity_analysis/SystematicSensitivityResult.java [hunk 4]
```diff
@@ -58,23 +59,25 @@
     private SensitivityComputationStatus status;
     private final StateResult nStateResult = new StateResult();
     private final Map<String, StateResult> contingencyResults = new HashMap<>();
+    private final Map<String, StateResult> contingencyResultsAfterCra = new HashMap<>();
```
Comment (alexandrem-rte) [inline sensitivity-analysis/src/main/java/com/farao_community/farao/sensitivity_analysis/SystematicSensitivityResult.java:62]: Could we rename this attribute "postCraResults" and the previous one "postContingencyResults"?

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/output/PreventiveAndCurativesRaoOutput.java [hunk 5]
```diff
@@ -221,7 +221,7 @@ public double getOptimizedSetPointOnState(State state, RangeAction rangeAction)
     @Override
     public Set<RangeAction> getActivatedRangeActionsDuringState(State state) {
         if (state.getInstant() == Instant.PREVENTIVE) {
-            return postPreventiveResult.getActivatedRangeActions().stream().filter(rangeAction -> isActivatedDuringState(state, rangeAction)).collect(Collectors.toSet());
+            return postPreventiveResult.getActivatedRangeActions();
         } else {
             return postCurativeResults.get(state).getActivatedRangeActions().stream().filter(rangeAction -> isActivatedDuringState(state, rangeAction)).collect(Collectors.toSet());
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/output/PreventiveAndCurativesRaoOutput.java:226]: true :)

File: sensitivity-analysis/src/main/java/com/farao_community/farao/sensitivity_analysis/SystematicSensitivityResult.java (hunk 4 above)
Comment (pet-mit) [reply sensitivity-analysis/src/main/java/com/farao_community/farao/sensitivity_analysis/SystematicSensitivityResult.java:62]: yes

**Review by alexandrem-rte — APPROVED** [review]

**Pull request merged**